README.md
[![License](https://img.shields.io/badge/license-MIT-green.svg?style=flat)](https://mit-license.org)
[![NPM](https://img.shields.io/npm/v/lumi-web-core.svg)](https://www.npmjs.org/package/lumi-web-core)

![lumicore](https://user-images.githubusercontent.com/63342220/80406279-7c663380-88cc-11ea-8b06-07825767b288.png)

# LumiCore
The LumiCore library is an implementation of tools for working with Bitcoin, Ethereum and Bitcoin Cash. It allows to create and work with mnemonic following the BIP39 standard, to run the private/public keys derivation tree following the BIP44 standard and sign transactions.

> Work of this module has been tested in applications at the Vue.js. You can find it [here](https://github.com/lumiwallet/lumi-web-core-app).

**Note**: Node version >= 10.19.0

## Installation
You can install the package via npm
``` console
npm install lumi-web-core
```
Or via yarn:
``` console
yarn add lumi-web-core
```

## Usage

> **Warning!** Do not use the mnemonic, keys, and addresses provided in this readme for real transactions!
>
### Create/import
Import the module and create a new class:
``` js
// typical import
import Wallet from 'lumi-web-core'

const WALLET = new Wallet()
```
You can create a new wallet by calling the method createNew(count). Count is the number of words for the new mnemonic.
By default it is 12. It can be also  12, 15, 18, 21, and 24 words.
``` js
// Create a new wallet
const CORE = await WALLET.createNew(15)
```
You can create a wallet based on an existing mnemonic, where mnemonic is a string consisting of words and spaces:
``` js
// Create wallet by new mnemonic
    const CORE = await WALLET.createByMnemonic('ramp grief anger ... olive main into throw')
```
Or xprv key: 
``` js
// Create wallet by xprv key
const CORE = await WALLET.createByKey('xprv9s21ZrQH143K49Xa838YbhNq...VtL5sLB9JsijRG51jHjuUSDVfDXPh5PFYNTqQ')
```
Example of the returned core:
``` js
{
    mnemonic: "ramp grief anger ... olive main into throw",
    xprv: "xprv9s21ZrQH143K49Xa838YbhNq...VtL5sLB9JsijRG51jHjuUSDVfDXPh5PFYNTqQ",
    from: "mnemonic",
    hdkey: Object,
    seed: Uint8Array(64),
    seedInHex: "1f53c65842ed3d0c54052f7f7315dbd9dcb...4af426ffb27234a0a571c44e29c1f4d1b181082e62d0a39",
    BTC: {
        address: "1MS1SjQ1...vrLhajoGLfiV"
        externalNode: Object,
        internalNode: Object,
    },
    ETH: {
        address: "0x06c019a17...aa6d949c947a02"
        node: Object,
        privateKey: Uint8Array(32),
        publicKey: Uint8Array(64),
    },
    BCH: {
        address: "1657JTP...5pFAA75HrD",
        externalNode:Object
        internalNode: Object
    }
}
```

### Derivation
You can get information about a child node using the method `getChildNodes`.
``` js
const data = {
  path: 'm/44'/0'/0'/0',
  from: 0,
  to: 20
}

const info = await WALLET.getChildNodes(data)
=> {
    node: {
        privateExtendedKey: "xprvA1CHKT9koKiKuhkeZ6XJxWXEQ1r...va4RtZfFmdYaQeNW72vX9q2DjMo3fsJwxCtkrVXyo2QP8",
        publicExtendedKey: "xpub6EBdixgedhGd8Bq7f84KKeTxx3gYWQ...bn9YVZ8PVXb6tTnCWTorLQPKqRggvdzyHH1TkoR7vuz"
    },
    list: [
        {
            path: "m/44'/0'/0'/0/0",
            privateKey: "KzwMNQ93Dt96Qg...mRpaBCmEXGH2Lpgr2dGZsV",
            publicKey: "023b693fa7fa22e505...4cc450a463c024ab1e3ec526ba",
            btcAddress: "1aVZBbZW...U6WvQF6w36H7",
            ethAddress: "0xcd1594ae...a74ceb13d889d",
            bchAddress: "bitcoincash:qzwpal7...c3djs884x56y9qptzl"
        },
        ...
    ]
}

```
`path` this is a string with a derivation path. The path should begin with `m/44'`.
`from` is top of the range derivation
`to` is end of the range derivation

### Creating a BTC transaction
To create a Bitcoin transaction you need to send a set of inputs and outputs to the `makeRawBtcTx` method:
``` js
const data = {
    inputs: [
        {
            addr: "1BYRALP...ypWzTQE8BkEL",
            value: 10000,
            n: 0,
            script: "76a91473a2...07d0bff2aee345ac88ac",
            hash: "de06df091735...afd145eecfcd5649634e1d5221",
            key: 'L3YFJ4cBDYhZ...bzqgSUxLjTAorR5Kc47mP6x1D'
        },
        {
            addr: "1BYRALPU...kLypWzTQE8BkEL",
            value: 10000,
            n: 0,
            script: "76a91473a2...7d0bff2aee345ac88ac",
            hash: "47e497fd8f6f4...5c2043e0a3c7ee0463cbc68e9",
            key: 'L3YFJ4cBDYhZ...xLjTAorR5Kc4AorR5K7mP6x1D'
        }
      ],
      outputs: [
        {
          address: '1NuABwx...5V5tiXpshAWb8W',
          value: 15000
        }
        {
          address: '1PbQ36GvG...nimPD7gHPQFk3Nu',
          value: 5000
        }
      ]
}

const btc_tx = await WALLET.makeRawBtcTx(data)
```
When the transaction is created successfully, an object with the transaction hash and raw tx data is returned
``` js
btc_tx => {
    hash: '6b4a00f9fd4e5c263b5af65fed0...14cafb97e0f5738cb4a8717',
    tx: '010000000221521d4e634956cdcfee45d1afbe79080a...04f3d76f2c65501976a914fbf7f38b23308227a308fffff88ac00000000'
}
```

### Creating an ETH transaction
To make a Ethereum transaction you need to fill in the following params:
``` js
const data = {
  nonce: 280,
  amount: 100000000000000,
  address: '0x1e8d99d2278...d89983e2920df33b485',
  gasPrice: 19950000002,
  gasLimit: 21000,
  privateKey: '0xd27c8544f946bd2a5456...d174e64e4a03030917bb8313'
}

const eth_tx = await WALLET.makeRawEthTx(data)
```

When the transaction is created successfully, an object with the transaction hash and raw tx data is returned
``` js
eth_tx => {
    hash: '0xcb3869c19f2462a572a5ff13...d67ca71c181452e52a698d035209',
    tx: '0xf86c8201188504a51cd782825208...6df4207768a01789b726296195a956ceb3dbe73798'
}
```

### Creating a BCH transaction
To create a Bitcoin Cash transaction you need to send a set of inputs and outputs to the `makeRawBchTx` method:
``` js
const data = {
    inputs: [
        {
            address: "bitcoincash:qzq29...rfavhxrhwvgm49vlhq",
            outputIndex: 1,
            satoshis: 604909,
            script: "76a91480a2a2...81a7acb98777388ac",
            txId: "0488b0c6678e527...5307ce7ddaa8cbc986e616",
            key: "KxMvzeEuGyrBRo...mfR9uzU6bAuc6kyXTmN8h"
        }
      ],
      outputs: [
        {
            address: "bitcoincash:qzdss...rfpnawutm9vlrh7wdcj",
            satoshis: 100000
        },
        {
            address: "bitcoincash:qr30f7...23hykq2fu3axuxec3g6c6",
            satoshis: 504231
        }
      ]
}

const bch_tx = await WALLET.makeRawBchTx(data)
```
Addresses included in inputs and outputs can be CashAddr format or Legacy format.
When the transaction is created successfully, an object with the transaction hash and raw tx data is returned
``` js
btc_tx => {
    hash: 'ac5cd881770c28aad990...5d181df4b7d5a9acfec3bdf',
    tx: '020000000116e686c9cba8da7dce075...edac9eff87677eccaa372580a4f23d3788ac00000000'
}
```

For more information, see the [docs](https://lumiwallet.github.io/lumi-web-core/).

## Nist testing
Testing [documentation](./nist/README.md).

## Created using
* [bip39](https://github.com/bitcoinjs/bip39)
* [bip39-checker](https://github.com/jcalfee/bip39-checker)
* [bitcoinjs-lib](https://github.com/bitcoinjs/bitcoinjs-lib)
* [ethereumjs-tx](https://github.com/ethereumjs/ethereumjs-vm/tree/master/packages/tx)
* [ethereumjs-util](https://github.com/ethereumjs/ethereumjs-util)
* [hdkey](https://github.com/cryptocoinjs/hdkey)
* [web3-utils](https://github.com/ethereum/web3.js)
* [wif](https://github.com/bitcoinjs/wif)
* [worker-loader](https://github.com/webpack-contrib/worker-loader)
* [bitcore-lib-cash](https://github.com/bitpay/bitcore/tree/master/packages/bitcore-lib-cash)
* [bchaddrjs](https://github.com/ealmansi/bchaddrjs)
## License

LumiCore is available under the MIT license. See the [LICENSE](LICENSE) file for more info.


CHANGELOG.md
- Added test for main core classes. Hotfix to get the last block of BTC
- Removed unused BTC and BCH core properties. Added new method
- Merge pull request #5 from lumiwallet/bch
- [UPDATE] doc, package version and changelog
- Edited a BCH address converter. Removed logs
- Added docs, edited bch tx methods
- Merge branch 'update-dependencies' into bch
- Added additional check for ETH api response
- Update @babel dependencies
- Added Bitcoin Cash support
- Merge branch 'master' into bch
- Merge pull request #4 from lumiwallet/fix-sync
- [UPDATE] version & docs
- Merge branch 'update-dependencies' into fix-sync
- Merge pull request #3 from lumiwallet/update-dependencies
- [UPDATE] package dependencies
- Added BCH core
- [FIX] btc sync and eth sync
- [UPDATE] npm dependencies
- [FIX] multi address request
- Merge pull request #2 from lumiwallet/dependabot/npm_and_yarn/elliptic-6.5.3
- Bump elliptic from 6.3.3 to 6.5.3
- [UPDATE] package-lock.json
- [UPDATE] docs and package version
- Merge pull request #1 from lumiwallet/edit-btc-fee
- Added prepare script
- [EDIT] Changed the generation of the list of commissions
- [ADD] npm shields
- [LUMICORE] v1.0.0


src/class/Core.js
import {validateMnemonic} from 'bip39'
import {normalize, checkWords} from 'bip39-checker'
import CustomError from '@/helpers/handleErrors'
import * as core from '@/helpers/coreHelper'

/**
 * Class Wallet
 * @class
 */

export default class Core {
  /**
   * Create a core
   * @param {Object} data
   * @param {string} data.from - A type new wallet creating: new, mnemonic or xprv
   * @param {number} data.count - Number of words for the new mnemonic. Is used when parameter from is 'new'
   * @param {string} data.mnemonic - The mnemonic phrase. It used when parameter from is 'mnemonic'
   * @param {string} data.key - BIP32 Root Key. It used when parameter from is 'mnemonic'
   */
  constructor (data = {}) {
    const {from, count, mnemonic, key} = data
    this.from = from
    this.count = count || 12
    this.mnemonic = mnemonic
    this.xprv = key
    this.seed = null
    this.hdkey = null
    this.BTC = {
      address: null,
      internalNode: null,
      externalNode: null
    }
    this.ETH = {
      address: null,
      node: null,
      privateKey: null,
      publicKey: null
    }
    this.BCH = {
      address: null,
      internalNode: null,
      externalNode: null
    }
    
    this.generateWallet()
  }
  
  /**
   * The main method that starts the generation of the core.
   * The type of generation depends on the parameter 'from'
   */
  
  generateWallet () {
    switch (this.from) {
      case 'new':
        this._generateNewMnemonic()
        break
      case 'mnemonic':
        this._importByMnemonic()
        break
      case 'xprv':
        this._importByKey()
        break
      default:
        this._generateNewMnemonic()
    }
    
    this._generateBTCcore()
    this._generateETHcore()
    this._generateBCHcore()
  }
  
  /**
   * Generating a new mnemonic.
   * The number of words in a mnemonic depends on the parameter 'count'
   * @private
   */
  
  _generateNewMnemonic () {
    const entropy = this._getEntropyLength(this.count)
    this.mnemonic = core.generateMnemonic(entropy)
    this.seed = core.mnemonicToSeed(this.mnemonic)
    this.hdkey = core.hdFromSeed(this.seed)
    this.xprv = core.getXprv(this.hdkey)
  }
  
  /**
   * Importing a wallet by mnemonic
   * @private
   */
  
  _importByMnemonic () {
    this.mnemonic = normalize(this.mnemonic)
    
    if (!this.checkMnemonic(this.mnemonic)) {
      throw new CustomError('err_core_mnemonic')
    }
    
    this.seed = core.mnemonicToSeed(this.mnemonic)
    this.hdkey = core.hdFromSeed(this.seed)
    this.xprv = core.getXprv(this.hdkey)
  }
  
  /**
   * Importing a wallet by key
   * @private
   */
  
  _importByKey () {
    if (!this.xprv || typeof this.xprv !== 'string') {
      throw new CustomError('err_core_xprv')
    }
    this.hdkey = core.hdFromXprv(this.xprv)
  }
  
  /**
   * Creating a core for Bitcoin.
   * At the output, we get a external and internal node
   * and the first address of the external core
   * @private
   */
  
  _generateBTCcore () {
    const bitcoin_external_path = 'm/44\'/0\'/0\'/0'
    const bitcoin_internal_path = 'm/44\'/0\'/0\'/1'
    this.BTC.externalNode = core.derive(this.hdkey, bitcoin_external_path)
    this.BTC.internalNode = core.derive(this.hdkey, bitcoin_internal_path)
    this.BTC.address = core.getBtcAddress(this.BTC.externalNode, 0)
  }
  
  /**
   * Creating a core for Ethereum.
   * At the output, we get a Ethereum node,
   * a private and public key, and the Ethereum address
   * @private
   */
  
  _generateETHcore () {
    const ethereum_path = 'm/44\'/60\'/0\'/0/0'
    this.ETH.node = core.derive(this.hdkey, ethereum_path)
    this.ETH.privateKey = core.getEthPrivateKey(this.ETH.node)
    this.ETH.privateKeyHex = '0x' + this.ETH.privateKey.toString('hex')
    this.ETH.publicKey = core.getEthPublicKey(this.ETH.privateKey)
    this.ETH.address = core.getEthAddress(this.ETH.publicKey)
  }
  
  /**
   * Creating a core for Bitcoin Cash.
   * At the output, we get a external and internal node
   * and the first address of the external core
   * @private
   */
  
  _generateBCHcore () {
    const bitcoincash_external_path = 'm/44\'/145\'/0\'/0'
    const bitcoincash_internal_path = 'm/44\'/145\'/0\'/1'
    this.BCH.externalNode = core.derive(this.hdkey, bitcoincash_external_path)
    this.BCH.internalNode = core.derive(this.hdkey, bitcoincash_internal_path)
    this.BCH.address = core.getBtcAddress(this.BCH.externalNode, 0)
  }
  
  /**
   * The method returns information about child nodes by the derivation path and range
   * @param {Object} data
   * @param {number} data.from - Top of the derivation range
   * @param {number} data.to - End of the derivation range
   * @param {string} data.path - Derivation path
   * @returns {{node: {privateExtendedKey: *, publicExtendedKey: *}, list: []}}
   */
  
  getChildNodes (data = {}) {
    let {from, to, path} = data
    
    from = +from
    to = +to
    
    if (!Number.isInteger(from) || !Number.isInteger(to) || from > to) {
      throw new CustomError('err_core_derivation_range')
    }
    
    try {
      const node = core.derive(this.hdkey, path)
      
      let info = {
        node: {
          privateExtendedKey: node.privateExtendedKey,
          publicExtendedKey: node.publicExtendedKey
        },
        list: []
      }
      
      for (let i = from; i <= to; i++) {
        const child = {}
        const deriveChild = node.deriveChild(i)
        child.path = `${ path }/${ i }`
        child.privateKey = core.privateKeyToWIF(deriveChild.privateKey)
        child.publicKey = deriveChild.publicKey.toString('hex')
        child.btcAddress = core.getBtcAddress(node, i)
        child.ethAddress = core.getEthAddressByNode(deriveChild)
        child.bchAddress = core.getCashAddress(child.btcAddress)
        info.list.push(child)
      }
      
      return info
    }
    catch (e) {
      throw new CustomError('err_core_derivation')
    }
  }
  
  /**
   * Checking the mnemonic for validity
   * @param mnemonic
   * @returns {boolean}
   */
  
  checkMnemonic (mnemonic) {
    let words = mnemonic.split(' ')
    let withTypo = []
    
    words.forEach((word, index) => {
      if (!checkWords(word, 'english')) {
        withTypo.push(index)
      }
    })
    
    if (withTypo.length) {
      throw new CustomError('err_core_mnemonic')
    }
    
    return validateMnemonic(mnemonic)
  }
  
  /**
   * Getting the entropy size by the number of words in a mnemonic
   * @param words
   * @returns {number} Bits of entropy
   * @private
   */
  
  _getEntropyLength (words) {
    let bitsOfEntropy = {
      12: 128,
      15: 160,
      18: 192,
      21: 224,
      24: 256
    }
    
    if (!bitsOfEntropy.hasOwnProperty(+words)) {
      throw new CustomError('err_core_7')
    }
    
    return bitsOfEntropy[words]
  }
  
  get DATA () {
    return {
      mnemonic: this.mnemonic,
      xprv: this.xprv,
      from: this.from,
      hdkey: this.hdkey,
      seed: this.seed,
      seedInHex: this.seed ? this.seed.toString('hex') : null,
      BTC: this.BTC,
      ETH: this.ETH,
      BCH: this.BCH
    }
  }
}
